### PR TITLE
[Issue 148] - Migrating from TSLint to ESLint

### PR DESCRIPTION
**Pull request by natesulat**
Signed-off-by: Nate Sulat <nates@kiva.org>

### CODE REVIEW

File: src/app/app.service.ts [hunk 1]
```diff
@@ -67,20 +70,20 @@ export class AppService {
      */
     public static async initAgent(app: INestApplication) {
         const agentService = await app.resolve(AgentService);
-        agentService.initProfilesFromDisk();
+        await agentService.initProfilesFromDisk();
```
Comment (matt-raffel-kiva) [inline src/app/app.service.ts:73]: This change concerns me because its been without the await operator for a long time.  maybe this was intentional to let it go run without waiting?

File: src/controller.handler/controller.handler.module.ts [hunk 2]
```diff
@@ -33,5 +32,6 @@ import { SecretsManager } from '../profile/secrets.manager';
                 CONTROLLER_HANDLER
             ],
         };
+        return Promise.resolve(dynamicModule);
```
Comment (matt-raffel-kiva) [inline src/controller.handler/controller.handler.module.ts:35]: This seems like the right change but (like above) I am concerned why it was in the code base for so long (and apparently working).

File: src/controller/agent.controller.service.ts [hunk 3]
```diff
@@ -47,7 +47,7 @@ export class AgentControllerService {
         if (process.env.MULTI_AGENT === 'true') {
             agentUrl = process.env.MULTITENANT_URL;
         } else {
-            const adminPort = ((profile && profile.adminApiPort) ? profile.adminApiPort : process.env.AGENT_ADMIN_PORT);
+            const adminPort: string = ((profile && profile.adminApiPort) ? profile.adminApiPort : process.env.AGENT_ADMIN_PORT);
```
Comment (matt-raffel-kiva) [inline src/controller/agent.controller.service.ts:50]: I need to be convinced its a good thing to remove type delcarations

File: src/controller.handler/single.controller.handler.ts [hunk 4]
```diff
@@ -51,6 +51,6 @@ export class SingleControllerHandler implements IControllerHandler {
      * For single controllers the api key is saved as an env var
      */
     public async handleAdminApiKey(): Promise<string> {
-        return process.env.ADMIN_API_KEY;
+        return Promise.resolve(process.env.ADMIN_API_KEY);
```
Comment (matt-raffel-kiva) [inline src/controller.handler/single.controller.handler.ts:54]: same

File: src/controller/handler/connections.ts [hunk 5]
```diff
@@ -111,3 +111,17 @@ export class Connections extends BaseAgentResponseHandler {
         return;
     }
 }
+
+interface Body {
```
Comment (matt-raffel-kiva) [inline src/controller/handler/connections.ts:115]: if we intend to have this interface, we need to put in its own file and follow our pattern for declaring interfaces

File: src/controller/handler/issue.credential.ts [hunk 6]
```diff
@@ -138,3 +138,26 @@ export class IssueCredential extends BaseAgentResponseHandler {
         return;
     }
 }
+
+interface Body {
```
Comment (matt-raffel-kiva) [inline src/controller/handler/issue.credential.ts:142]: same.  I do not understand why we are creating interfaces here.

**Review by matt-raffel-kiva — CHANGES_REQUESTED** [review]

File: src/app/app.service.ts (hunk 1 above)
Comment (natesulat) [reply src/app/app.service.ts:73]: Could be. There are enough `require-await` linting errors in our codebase that we might want to think about just turning it off by default across our apps.

File: src/app/app.service.ts [hunk 7]
```diff
@@ -67,20 +70,20 @@ export class AppService {
...
             await agentService.init();
         } catch (e) {
-            Logger.log(`Failed to start agent, retrying... ${e.message}`, e);
+            Logger.log(`Failed to start agent, retrying... ${e.message as string}`, e);
             try {
                 await ProtocolUtility.delay(1000);
                 await agentService.init();
             } catch (e2) {
-                Logger.log(`Failed to start agent, exiting... ${e2.message}`, e2);
+                Logger.log(`Failed to start agent, exiting... ${e2.message as string}`, e2);
```
[the remark this answers is not in the document]
Comment (natesulat) [reply src/app/app.service.ts:86]: The linter barks because `e2` implicitly has an `any` type, which means that `e2.message` also implicitly has an `any` type. There's a rule that you can only use certain types in string interpolation.

Given that the transpiler will remove this, I'm not terribly worried about the effect on our packages. We can discuss further, though.

File: src/caller/caller.interface.ts [hunk 8]
```diff
@@ -5,15 +5,15 @@ export interface ICaller {
     /**
      * Spins up a new agent, specific implementations handle single vs multi agent
      */
-    spinUpAgent(): Promise<any>;
+    spinUpAgent(): Promise<any>
```
[the remark this answers is not in the document]
Comment (natesulat) [reply src/caller/caller.interface.ts:8]: Explained in this comment: https://github.com/kiva/aries-key-guardian/pull/107#discussion_r841830666

TL;DR, the autogenerated rules leans towards not using extra characters if you're separating interface members with a newline.

File: src/controller/agent.controller.service.ts (hunk 3 above)
Comment (natesulat) [reply src/controller/agent.controller.service.ts:50]: We're adding a type declaration here, though.

Comment (matt-raffel-kiva) [reply src/controller/agent.controller.service.ts:50]: agreed.  I just want to have the conversation in general.  Ill mark this as resolved, specifically.

File: .eslintignore [hunk 9]
```diff
@@ -0,0 +1,2 @@
+test
```
Comment (jeffk-kiva) [inline .eslintignore:1]: We shouldn't be ignoring the test directory.

File: src/controller.handler/controller.handler.module.ts (hunk 2 above)
Comment (jeffk-kiva) [reply src/controller.handler/controller.handler.module.ts:35]: My worried thought is that the tests were maybe passing trivially. Tests passing != everything is working.

File: src/controller/agent.governance.ts [hunk 10]
```diff
@@ -145,8 +145,8 @@ export class AgentGovernance {
      */
     public async invokeHandler(agentUrl: string, agentId: string, adminApiKey: string, route: string, topic: string,
                                body: any, token?: string) : Promise<any> {
-        let result: boolean = false;
-        // tslint:disable-next-line:forin
+        let result = false;
+        // eslint-disable-next-line @typescript-eslint/no-for-in-array
```
Comment (jeffk-kiva) [inline src/controller/agent.governance.ts:149]: Can we delete this rule? I don't understand why it doesn't like `for in` arrays, they seem perfectly fine to me.

File: src/controller/handler/connections.ts (hunk 5 above)
Comment (jeffk-kiva) [reply src/controller/handler/connections.ts:115]: @matt-raffel-kiva TBH this repo probably needs a good deal of refactoring. I'm with you on pulling this out though. In general, I would like to see **a lot** more use of types and type validations in this repo.

File: src/controller/handler/issue.credential.ts (hunk 6 above)
Comment (jeffk-kiva) [reply src/controller/handler/issue.credential.ts:142]: Yeah, it might come from an external source, but that external source needs to obey the contract defined by aries-controller. Let's pull this out as a DTO and add some validations to it.

File: src/issuer/issuer.service.ts [hunk 11]
```diff
@@ -61,9 +61,9 @@ export class IssuerService {
     /**
      * Record the issued credential using the external credential record service
      */
-    public async recordCredential(entityData: any, issuanceRecord: any) {
+    public async recordCredential(entityData: any, issuanceRecord: any): Promise<void> {
         const url = process.env.CREDENTIAL_RECORD_URL;
-        this.callService('POST', url, {
+        await this.callService('POST', url, {
```
Comment (jeffk-kiva) [inline src/issuer/issuer.service.ts:66]: I'm not sure if it's safe to `await` here, it might be intentionally asynchronous. @matt-raffel-kiva do you know?

File: src/controller/handler/connections.ts (hunk 5 above)
Comment (matt-raffel-kiva) [reply src/controller/handler/connections.ts:115]: Agreed.  The challenge comes from Acapy which is calling us with the data structures and it is not easily identifiably consistent as the structures vary depending on factors like where the message originates.  I will write up an architectural issue for us to work on.

**Review by jeffk-kiva — APPROVED** [review]

File: package.json [hunk 12]
```diff
@@ -31,8 +31,8 @@
     "prestart:prod": "rimraf dist && tsc",
```
Comment (matt-raffel-kiva) [inline package.json:31]: make sure to bump the package version of this package since thats how we push updates downstream

**Review by matt-raffel-kiva — CHANGES_REQUESTED** [review]

File: package.json (hunk 12 above)
Comment (natesulat) [reply package.json:31]: Whoops. Thanks!
